### PR TITLE
models: add workspace retention rules table

### PR DESCRIPTION
**Pull request by audrium**
closes https://github.com/reanahub/reana-db/issues/172

To test:

- `kubectl exec -i -t deployment/reana-server -c rest-api -- invenio shell`:
```
from reana_db.models import Workflow, WorkspaceRetentionRule, WorkspaceRetentionRuleStatus
from reana_db.utils import update_workspace_retention_rules
from reana_db.database import Session
from datetime import datetime
s = Session()

workflow_id = Workflow.query.first().id_
workspace_files = "myfetched/**/*.root"
retention_days = 1
apply_on = None

# Insert a single rule
wr = WorkspaceRetentionRule(workflow_id=workflow_id, workspace_files=workspace_files, retention_days=retention_days)
s.add(wr)
s.commit()

# Update a single rule
wr.status = WorkspaceRetentionRuleStatus.inactive
s.add(wr)
s.commit()

# Inspect audit logs
wr.audit_logs.all()

# Insert multiple rules while creating the workflow
w = Workflow.query.first()
retention_rules = [
    {
        "workspace_files": "myfiltered/*.root",
        "retention_days": 7
    },
    {
        "workspace_files": "myfits/",
        "retention_days": 30
    },
]
w.set_workspace_retention_rules(retention_rules)

# Update multiple rules
wr_all = WorkspaceRetentionRule.query.filter(WorkspaceRetentionRule.workflow_id == workflow_id).all()
for rule in wr_all:
    rule.status = WorkspaceRetentionRuleStatus.applied
    s.add(rule)
s.commit()

# Update all workflow rules
workflow = Workflow.query.first()
workflow.inactivate_workspace_retention_rules()
workflow.activate_workspace_retention_rules()

# Try to transit to a wrong status
update_workspace_retention_rules(workflow, WorkspaceRetentionRuleStatus.created)
```

### CODE REVIEW

File: reana_db/models.py [hunk 1]
```diff
@@ -812,6 +833,73 @@ def __repr__(self):
...
+    action = Column(
+        Enum(WorkspaceRetentionRuleStatus), primary_key=True, nullable=False
+    )
+
+    def __repr__(self):
+        """Workspace Retention audit log string representation."""
+        return f"<WorkspaceRetentionAuditLog {self.workspace_retention_rule_id} {self.action}>"
+
+
+class WorkspaceRetention(Base):
```
Comment (tiborsimko) [inline reana_db/models.py:867]: We have three new classes with this PR:

- class WorkspaceRetentionRuleStatus(enum.Enum):
- class WorkspaceRetentionAuditLog(Base):
- class WorkspaceRetention(Base):

Quick impressions seeing these names: the first is nicely self-understandable, the third seem to be unfinished (retention what?) as if calling for a final noun (WorkspaceRetentionRule?), and the second is perhaps too businessy if we include "audit" (WorkspaceRetentionLogs?) and may perhaps clash with the AuditLog table that we have already for some user actions. 

(BTW it may be a good idea to reuse the existing AuditLog one, so that we have centralised all the "auditable actions" in one place... but that would need more changes.)

File: reana_db/models.py [hunk 2]
```diff
@@ -627,6 +633,42 @@ def get_workspace_disk_usage(self, summarize=False, search=None):
 
         return [{"name": "", "size": disk_usage}]
 
+    def set_workspace_retention_rules(self, rules: List[Dict[str, str]]):
+        """Set workspace retention rules for the workflow."""
+        from .database import Session
+
+        for rule in rules:
```
Comment (mdonadoni) [inline reana_db/models.py:640]: What do you think about having a limit on the number of rules? (in any case, we can do this in another PR)

File: reana_db/utils.py [hunk 3]
```diff
@@ -351,6 +352,35 @@ def update_users_cpu_quota(user=None) -> None:
...
+            continue
+        if not rule.can_transition_to(status):
+            raise Exception(
+                f"Cannot transition workspace retention rule {rule.id_} "
+                f"from status {rule.status} to {status}."
+            )
+        if status == WorkspaceRetentionRuleStatus.active:
+            rule.apply_on = datetime.today().replace(
+                hour=23, minute=59, second=59
+            ) + timedelta(days=rule.retention_days)
```
Comment (mdonadoni) [inline reana_db/utils.py:378]: Should we also set `apply_on` to `None` when inactivating rules?

**Review by mdonadoni — COMMENTED** [review]
The PR works well, good job! I have left just a few comments

**Comment by mdonadoni** [comment]
Late comment: I think it would be a good idea to add a few tests to this PR!

File: reana_db/models.py (hunk 2 above)
Comment (audrium) [reply reana_db/models.py:640]: I think it could be tackled as part of [validation issue](https://github.com/reanahub/reana-server/issues/507). I've linked this comment there in the description

File: reana_db/models.py [hunk 4]
```diff
@@ -812,6 +856,106 @@ def __repr__(self):
...
+        "WorkspaceRetentionAuditLog", backref="workspace_retention_rule", lazy="dynamic"
+    )
+
+    def can_transition_to(self, next_status):
+        """Whether the provided retention rule can transition to the next status."""
+        current_transition = (self.status, next_status)
+        return current_transition in ALLOWED_WORKSPACE_RETENTION_RULE_STATUS_TRANSITIONS
+
+    @property
+    def serialize(self):
```
[the remark this answers is not in the document]
Comment (mdonadoni) [reply reana_db/models.py]: > Do you have something specific in mind? Probably better to create an issue about this

I was not thinking about anything in particular, we can create an issue to investigate how to proceed!

**Review by mdonadoni — APPROVED** [review]
Works well, great job! :rocket:
